tests/integration/test_using_cached_dependencies.py

```diff
@@ -382,8 +389,77 @@ def update_expected_data(env_data, replace_rules):
+        env_data["content_manifest"]["purl"], replace_rules
+    )
+    for i, p in enumerate(env_data["content_manifest"]["dep_purls"]):
+        env_data["content_manifest"]["dep_purls"][i] = replace_by_rules(p, replace_rules)
+    for i, p in enumerate(env_data["content_manifest"]["source_purls"]):
+        env_data["content_manifest"]["source_purls"][i] = replace_by_rules(p, replace_rules)
+
+
+def update_main_repo(env_data, repo_dir, tmpdir, new_dep_commits, dep_repo):
+    """
+    Update main repo with new dependencies and return replacement rules.
+
+    Changes depend on package manager:
+    pip:
+        * download dep archive and calculate a hash
+        * update requirements.txt with 2 deps (https + VCS)
+        * return replacement rules based on commits and hash
+
+    gomod:
+        * get a pseudo-version for dependency commit
+        * replace a `require` string with old dependency
+          with new one in the go.mod file
+        * return replacement rules based on commit and pseudo-version
+
+    :param dict env_data: the test data
+    :param str repo_dir: path to the main repository
+    :param tmpdir: temporary test directory
+    :param list new_dep_commits: list of 2 dep commits
+    :param dep_repo: Dependency git repository
+    :return: replacement rules
+    :rtype: dict
+    """
+    if env_data["pkg_managers"] == ["pip"]:
+        # Download the archive with first commit changes
+        archive_name = os.path.join(tmpdir, f"{new_dep_commits[0]}.zip")
+        utils.download_archive(
+            f"{env_data['dep_archive_baseurl']}{new_dep_commits[0]}.zip", archive_name
+        )
+        # Get the archive hash
+        dep_hash = utils.get_sha256_hash_from_file(archive_name)
+
+        # Add new dependencies into the main repo
+        with open(os.path.join(repo_dir, "requirements.txt"), "a") as f:
+            f.write(
+                f"{env_data['dep_archive_baseurl']}{new_dep_commits[0]}"
+                f".zip#egg=appr&cachito_hash=sha256:{dep_hash}\n"
+            )
+            f.write(f"git+{env_data['https_dep_repo']}@{new_dep_commits[1]}#egg=appr\n")
+
+        # return replacement rules
+        return {
+            "FIRST_DEP_COMMIT": new_dep_commits[0],
+            "SECOND_DEP_COMMIT": new_dep_commits[1],
+            "FIRST_DEP_HASH": dep_hash,
+        }
+    elif env_data["pkg_managers"] == ["gomod"]:
+        dep_version = get_pseudo_version(dep_repo, new_dep_commits[0])
+
+        with open(os.path.join(repo_dir, "go.mod"), "r") as f:
+            base_gomod_file = [
+                line for line in f.read().split("\n") if not line.startswith("require")
+            ]
+        with open(os.path.join(repo_dir, "go.mod"), "w+") as f:
+            for line in base_gomod_file:
+                f.write(f"{line}\n")
+            go_dep = env_data["https_dep_repo"][
+                len("https://") : env_data["https_dep_repo"].find(".git")
```

Comment:
This could break if the url does not end with `.git`, could you fix this or add a comment to the test data (to say that the url must end with `.git`)?

Reply:
@chmeliik Yes, you are right! Let's avoid special rules